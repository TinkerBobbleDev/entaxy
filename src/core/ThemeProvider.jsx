import React from 'react'
import PropTypes from 'prop-types'
import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles'
import CssBaseline from '@material-ui/core/CssBaseline'
import blueGrey from '@material-ui/core/colors/blueGrey'
import '../styles/globals.css'

const theme = createMuiTheme({
  palette: {
    primary: { main: blueGrey[800] },
    secondary: { main: '#1B9CC4' },
    info: {
      text: '#00acc1',
      background: '#00acc11a'
    },
    danger: {
      text: '#721c24',
      background: '#f8d7da66'
    }

  },
  typography: {
    useNextVariants: true
  }
})

const ThemeProvider = ({ children }) => (
  <CssBaseline>
    <MuiThemeProvider theme={theme}>
      {children}
    </MuiThemeProvider>
  </CssBaseline>
)

ThemeProvider.propTypes = {
  children: PropTypes.node.isRequired
}

export default ThemeProvider
