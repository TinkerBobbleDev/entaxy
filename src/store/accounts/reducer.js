/* eslint-disable no-case-declarations */
import types from './types'

export const initialState = {
  byId: {}, // {id1: {id: id1, name: 'account1', institution: 'TD'}, id2: {...}}
  byInstitution: {} // {TD: {accountIds: [id1, id2], balance: 100, apiKey: XYZ}, BMO: {...}}
}

export const groupByInstitution = ({ byId, byInstitution }) => {
  // Group sorted accounts by institution
  // results in something like:
  // {
  //   TD: {accountIds: [id1, id2], balance: 100},
  //   BMO: {accountIds: [id3], balance: 200}
  // }
  const groupedAccountIds = Object.values(byId)
    // Sort accounts by name
    .sort((a, b) => (a.name > b.name))
    // Group them by institution
    .reduce((result, account) => {
      const {
        id,
        institution,
        groupId,
        currentBalance
      } = account

      // Merge existing institution data
      const existingInstitutionData = byInstitution[institution] || { groups: {} }
      const currentInstitutionData = (result[institution] || { balance: 0, groups: {} })
      const newInstitutionData = { ...existingInstitutionData, ...currentInstitutionData }

      // Merge existing group data
      const existingGroupData = existingInstitutionData.groups[groupId] || {}
      const currentGroupData = currentInstitutionData.groups[groupId] || {
        id: groupId,
        balance: 0,
        accountIds: []
      }
      const newGroupData = { ...existingGroupData, ...currentGroupData }

      return {
        ...result,
        [institution]: {
          ...newInstitutionData,
          balance: currentInstitutionData.balance + currentBalance.localCurrency,
          groups: {
            ...newInstitutionData.groups,
            [groupId]: {
              type: 'default',
              ...newGroupData,
              balance: currentGroupData.balance + currentBalance.localCurrency,
              accountIds: [...newGroupData.accountIds, id]
            }
          }
        }
      }
    }, {})

  // But now the institutions are not sorted and we lost existing institution data
  return Object.keys(groupedAccountIds)
    // Sort institutions by name
    .sort((a, b) => (a > b))
    // Merge new sorted institutions with existing institution data
    .reduce((result, institution) => ({
      ...result,
      [institution]: {
        ...byInstitution[institution],
        ...groupedAccountIds[institution]
      }
    }), {})
}

let accounts
let institutionData
export default (state = initialState, { type, payload }) => {
  switch (type) {
    case types.LOAD_ACCOUNTS:
      return payload || initialState
    case types.CREATE_ACCOUNT:
      accounts = { ...state.byId, [payload.id]: payload }
      return {
        ...state,
        byId: accounts
      }
    case types.UPDATE_ACCOUNT:
      accounts = { ...state.byId, [payload.id]: payload }
      return {
        ...state,
        byId: accounts
      }
    case types.DELETE_ACCOUNT:
      const { [payload]: _, ...rest } = state.byId
      return {
        ...state,
        byId: rest
      }
    case types.GROUP_BY_INSTITUTION:
      return {
        ...state,
        byInstitution: groupByInstitution(state)
      }
    case types.CREATE_ACCOUNT_GROUP:
    case types.UPDATE_ACCOUNT_GROUP:
      if (payload === undefined) return state
      const { institution, accountGroup } = payload
      institutionData = state.byInstitution[institution] || { groups: {} }
      return {
        ...state,
        byInstitution: {
          ...state.byInstitution,
          [institution]: {
            ...institutionData,
            groups: {
              ...institutionData.groups,
              [accountGroup.id]: accountGroup
            }
          }
        }
      }
    default:
      return state
  }
}
