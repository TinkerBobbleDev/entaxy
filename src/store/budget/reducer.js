import types from './types'
import budgetCategories from '../../data/budgetCategories'

// https://projects.susielu.com
// http://repec.sowi.unibe.ch/stata/palettes/index.html
const defaultColours = [
  '#1f78b4', '#b2df8a', '#e31a1c',
  '#ff7f00', '#cab2d6', '#a6cee3',
  '#33a02c', '#6a3d9a', '#fb9a99',
  '#fdbf6f', '#ffff99', '#b15928'
]

export const initialState = (() => {
  const colours = {}
  let count = 0
  const categories = Object.keys(budgetCategories).sort().map(category => (
    {
      label: category,
      options: budgetCategories[category].map((subCategory) => {
        const colour = defaultColours[count % defaultColours.length]
        colours[subCategory] = colour
        count += 1
        return { label: subCategory, value: subCategory, colour }
      })
    }
  ))
  return {
    categories,
    colours,
    rules: {} // rules have the format => match: { category: 'cat 1', type: 'exact_match', count: 0 }
  }
})()

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_BUDGET:
      return action.payload || initialState
    case types.CREATE_EXACT_RULE:
      return {
        ...state,
        rules: {
          ...state.rules,
          [action.payload.match]: {
            category: action.payload.category,
            type: 'exact_match',
            count: 0
          }
        }
      }
    case types.DELETE_EXACT_RULE:
      // eslint-disable-next-line no-case-declarations
      const r = Object.keys(state.rules).reduce((result, match) => {
        if (match !== action.payload) {
          return {
            ...result,
            rules: { ...result.rules, [match]: state.rules[match] }
          }
        }
        return result
      }, { ...state, rules: {} })
      return r
    case types.COUNT_RULE_USAGE:
      // reset counters
      // eslint-disable-next-line no-case-declarations
      const newState = Object.keys(state.rules).reduce((result, match) => (
        {
          ...result,
          rules: {
            ...result.rules,
            [match]: { ...result.rules[match], count: 0 }
          }
        }
      ), state)

      // count the transactions
      for (const transaction of action.payload) {
        if (transaction.description in newState.rules) {
          newState.rules[transaction.description].count += 1
        }
      }
      // remove unused rules
      return Object.keys(newState.rules).reduce((result, match) => {
        if (newState.rules[match].count > 0) {
          return {
            ...result,
            rules: { ...result.rules, [match]: newState.rules[match] }
          }
        }
        return result
      }, { ...newState, rules: {} })
    default:
      return state
  }
}
