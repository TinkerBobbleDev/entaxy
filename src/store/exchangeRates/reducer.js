import { omit } from 'lodash'
import parse from 'date-fns/parse'
import types from './types'

export const initialState = {}

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_EXCHANGE_RATES:
      return action.payload || initialState
    case types.UPDATE_EXCHANGE_RATES:
      // payload = {
      //   '2019-01-01': { GBP: 1.4, CAD: 1.0 },
      //   '2019-02-02': { GBP: 1.3}
      // }
      // eslint-disable-next-line no-case-declarations
      const newState = Object.keys(action.payload)
        .reduce((res1, date) => {
          return Object.keys(action.payload[date])
            .reduce((res2, currency) => ({
              ...res2,
              [currency]: {
                ...res2[currency],
                [parse(date).getTime()]: action.payload[date][currency]
              }
            }), res1)
        }, state)
      Object.keys(newState).forEach((currency) => {
        newState[currency].dates = Object.keys(newState[currency])
          .filter(c => c !== 'dates')
          .sort((a, b) => b - a)
      })
      return newState
    case types.DELETE_CURRENCIES:
      return omit(state, action.payload)
    default:
      return state
  }
}
