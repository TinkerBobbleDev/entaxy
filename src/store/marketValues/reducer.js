import _ from 'lodash'
import types from './types'

export const initialState = {}

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_MARKET_VALUES:
      return action.payload || initialState
    case types.CREATE_MARKET_VALUE:
      return {
        ...state,
        [action.payload.ticker]: {
          marketValue: action.payload.marketValue,
          updatedOn: action.payload.updatedOn
        }
      }
    case types.UPDATE_MARKET_VALUE:
      return {
        ...state,
        [action.payload.ticker]: {
          marketValue: action.payload.marketValue,
          updatedOn: action.payload.updatedOn
        }
      }
    case types.DELETE_MARKET_VALUES:
      return _.omit(state, action.payload)
    default:
      return state
  }
}
