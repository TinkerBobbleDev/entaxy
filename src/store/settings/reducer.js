import _ from 'lodash'
import types from './types'

export const initialState = {
  portfolioFilters: {
    source: {},
    account: {}
  }
}

export default (state = initialState, action) => {
  switch (action.type) {
    case types.LOAD_SETTINGS:
      return action.payload || initialState
    case types.UPDATE_PORTFOLIO_FILTER_VALUE:
      return {
        ...state,
        portfolioFilters: {
          ...state.portfolioFilters,
          [action.payload.filterName]: {
            ...state.portfolioFilters[action.payload.filterName],
            [action.payload.option]: action.payload.value
          }
        }
      }
    case types.CREATE_PORTFOLIO_FILTERS:
      return {
        ...state,
        portfolioFilters: {
          ...state.portfolioFilters,
          [action.payload.filterName]: {
            ...state.portfolioFilters[action.payload.filterName],
            ...action.payload.options
          }
        }
      }
    case types.DELETE_PORTFOLIO_FILTERS:
      return {
        ...state,
        portfolioFilters: {
          ...state.portfolioFilters,
          [action.payload.filterName]: _.omit(state.portfolioFilters[action.payload.filterName], action.payload.options)
        }
      }
    default:
      return state
  }
}
