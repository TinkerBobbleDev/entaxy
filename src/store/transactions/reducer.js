import _ from 'lodash'
import types from './types'

export const initialState = []

export default (state = initialState, action) => {
  let index = null

  switch (action.type) {
    case types.LOAD_TRANSACTIONS:
      return action.payload || initialState
    case types.CREATE_TRANSACTION:
      return [...state, action.payload]
    case types.DELETE_TRANSACTION:
      index = _.findIndex(state, transaction => transaction.id === action.payload)
      return [...state.slice(0, index), ...state.slice(index + 1)]
    default:
      return state
  }
}
