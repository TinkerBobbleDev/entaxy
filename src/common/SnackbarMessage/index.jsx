import React from 'react'
import PropTypes from 'prop-types'
import { compose } from 'recompose'
import { connect } from 'react-redux'
import { withStyles } from '@material-ui/core/styles'
import Snackbar from '@material-ui/core/Snackbar'
import CheckCircleIcon from '@material-ui/icons/CheckCircle'
import ErrorIcon from '@material-ui/icons/Error'
import { hideSnackbar } from '../../store/settings/actions'

const styles = (theme) => ({
  message: {
    display: 'flex',
    alignItems: 'center'
  },
  icon: {
    fontSize: 20,
    opacity: 0.9,
    marginRight: theme.spacing(2)
  },
  success: {
    color: theme.palette.success.icon
  },
  error: {
    color: theme.palette.danger.icon
  }
})

const mapStateToProps = ({ settings }) => {
  return { settings }
}

const mapDispatchToProps = (dispatch) => ({
  handleCloseSnackbar: () => { dispatch(hideSnackbar()) }
})

const SnackbarMessage = ({ classes, settings, handleCloseSnackbar }) => (
  <Snackbar
    anchorOrigin={{ vertical: 'bottom', horizontal: 'left' }}
    open={settings.snackbarMessage !== null}
    onClose={handleCloseSnackbar}
    autoHideDuration={4000}
    ContentProps={{ 'aria-describedby': 'message-id' }}
    message={(
      <span id="message-id" className={classes.message}>
        {settings.snackbarMessage && settings.snackbarMessage.status === 'success' && (
          <CheckCircleIcon className={[classes.icon, classes.success].join(' ')} />
        )}
        {settings.snackbarMessage && settings.snackbarMessage.status === 'error' && (
          <ErrorIcon className={[classes.icon, classes.error].join(' ')} />
        )}
        {settings.snackbarMessage && settings.snackbarMessage.text}
      </span>
    )}
  />
)
SnackbarMessage.propTypes = {
  classes: PropTypes.object.isRequired,
  settings: PropTypes.object.isRequired,
  handleCloseSnackbar: PropTypes.func.isRequired
}

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withStyles(styles)
)(SnackbarMessage)
