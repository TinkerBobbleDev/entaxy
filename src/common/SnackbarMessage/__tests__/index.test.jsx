import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import ThemeProvider from '../../../core/ThemeProvider'
import { initialState as userInitialState } from '../../../store/user/reducer'

import SnackbarMessage from '..'

describe('SnackbarMessage', () => {
  it('it doesn\'t render anything if there\'s no message', () => {
    const mockStore = configureMockStore()
    const store = mockStore({ user: userInitialState })
    const { queryByTestId } = render(
      <Provider store={store}>
        <ThemeProvider>
          <SnackbarMessage />
        </ThemeProvider>
      </Provider>
    )
    expect(queryByTestId('snackback')).not.toBeInTheDocument()
  })

  it('it renders a success message correctly', () => {
    const mockStore = configureMockStore()
    const store = mockStore({ user: { snackbarMessage: { text: 'message', status: 'success' } } })
    const { getByText, queryByText, getByTestId } = render(
      <Provider store={store}>
        <ThemeProvider>
          <SnackbarMessage />
        </ThemeProvider>
      </Provider>
    )
    expect(getByTestId('snackback')).toBeInTheDocument()
    expect(getByText('message')).toBeInTheDocument()
    expect(getByTestId('successIcon')).toBeInTheDocument()
    expect(queryByText('errorIcon')).not.toBeInTheDocument()
    // The snackbar should disapear in 4 secs so we click the body to make it isaoear sooner
    fireEvent.click(getByText((_, element) => element.tagName === 'BODY'))
    expect(store.getActions()).toEqual([{ type: 'HIDE_SNACKBAR' }])
  })

  it('it renders a error message correctly', () => {
    const mockStore = configureMockStore()
    const store = mockStore({ user: { snackbarMessage: { text: 'message', status: 'error' } } })
    const { getByText, queryByText, getByTestId } = render(
      <Provider store={store}>
        <ThemeProvider>
          <SnackbarMessage />
        </ThemeProvider>
      </Provider>
    )
    expect(getByTestId('snackback')).toBeInTheDocument()
    expect(getByText('message')).toBeInTheDocument()
    expect(getByTestId('errorIcon')).toBeInTheDocument()
    expect(queryByText('successIcon')).not.toBeInTheDocument()
    // The snackbar should disapear in 4 secs so we click the body to make it isaoear sooner
    fireEvent.click(getByText((_, element) => element.tagName === 'BODY'))
    expect(store.getActions()).toEqual([{ type: 'HIDE_SNACKBAR' }])
  })
})
