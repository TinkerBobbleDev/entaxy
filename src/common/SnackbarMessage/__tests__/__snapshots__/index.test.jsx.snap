// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnackbarMessage matches snapshot 1`] = `null`;
